src/utils/transaction/serializeTransaction.ts

```
@@ -0,0 +1,222 @@
+    gas ? toHex(gas) : '0x',
+    encodedTo,
+    value ? toHex(value) : '0x',
+    data ?? '0x',
+    encondedAccessList,
+  ]
+
+  if (signature) {
+    if (!(isHash(signature.s) && isHash(signature.r))) {
+      throw new InvalidHashError({ hash: [signature.s, signature.r] })
+    }
+    encodedTransaction.push(
+      signature.v === 27n ? '0x' : toHex(1),
+      signature.r,
+      signature.s,
+    )
+  }
+
+  return concatHex(['0x01', toRlp(encodedTransaction)]) as EIP2930Serialized
+}
+
+function serializeTransactionLegacy(
+  transaction: Omit<TransactionRequestLegacy, 'from'> & { chainId: number },
+  signature?: Signature,
+): Hex {
+  const { chainId, gas, data, nonce, to, value, gasPrice } = transaction
+
+  assertTransactionNonEIP1559('legacy', transaction)
+
+  const encodedTo = to ?? '0x'
+
+  const encodedTransaction = [
+    nonce ? toHex(nonce) : '0x',
+    gasPrice ? toHex(gasPrice) : '0x',
+    gas ? toHex(gas) : '0x',
+    encodedTo,
+    value ? toHex(value) : '0x',
+    data ?? '0x',
+  ]
+
+  if (signature) {
+    return toRlp([
+      ...encodedTransaction,
+      toHex(BigInt(chainId * 2) + BigInt(35n + signature.v - 27n)),
+      signature.r,
+      signature.s,
+    ])
+  }
+
+  return toRlp([...encodedTransaction, toHex(chainId), '0x', '0x'])
+}
+
+export function serializeTransaction<
+  TTransactionType extends Omit<TransactionRequest, 'from'> & {
+    chainId: number
+  },
+  TOptions extends {
+    type: TTransactionType['maxPriorityFeePerGas'] extends bigint
+      ? 'eip1559'
+      : TTransactionType['maxFeePerGas'] extends bigint
+      ? 'eip1559'
+      : TTransactionType['gasPrice'] extends bigint
+      ? TTransactionType['accessList'] extends AccessList
+        ? 'eip2930'
+        : 'legacy'
+      : TransactionType
+    signature?: Signature
+  },
+>(
+  transactionType: TTransactionType,
+  options: TOptions,
+): SerializedType<TOptions['type']> {
+  if (options.type === 'eip1559') {
```

Yeah, so related to my previous comment, I think we should have a function like `getTransactionType(transaction)` where we can retrieve the txn type instead of the consumer defining it.

pseudo:

```ts
const type = getTransactionType(transaction)

if (type === 'eip1559') return serializeTransactionEIP1559(transaction)
if (type === 'eip2930') return serializeTransactionEIP2930(transaction)
return serializeTransactionLegacy(transaction)
```

---

I’m okay with having this as an option, however, I think we still may need to consider the case where we can’t derive the type statically. Maybe if no type is passed through, we try to derive it from the transaction properties? What do you think?

---

I see your point. I'm fine with either option honestly. The only question that I have in infering the type from the transaction object is, what would be the expected behaviour if someone passes something like `{ chainId: 1 }` as a argument and none of the key props are available to infer the transaction type? In this case we would need to define a fallback type or we could throw an error. What do you think?

---

I'm comfortable with throwing an error if a type cannot be derived. If folks want to construct the most minimal transaction signatures, they will need to at least provide a type.